control-plane/agents/src/bin/core/controller/reconciler/node/snapshot.rs

```diff
@@ -0,0 +1,96 @@
+use crate::controller::{
+    io_engine::ReplicaSnapshotApi,
+    reconciler::poller::ReconcilerWorker,
+    resources::{OperationGuardArc, ResourceUid},
+    task_poller::{PollContext, PollResult, PollTimer, PollerState, TaskPoller},
+};
+use stor_port::types::v0::{
+    store::node::NodeSpec,
+    transport::{DestroyReplicaSnapshot, ListReplicaSnapshots, SnapshotId},
+};
+
+/// Node snapshot garbage collector.
+#[derive(Debug)]
+pub(super) struct NodeSnapshotGarbageCollector {
+    counter: PollTimer,
+}
+impl NodeSnapshotGarbageCollector {
+    /// Return a new `Self`.
+    pub(super) fn new() -> Self {
+        Self {
+            counter: ReconcilerWorker::garbage_collection_period(),
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl TaskPoller for NodeSnapshotGarbageCollector {
+    async fn poll(&mut self, context: &PollContext) -> PollResult {
+        let nodes = context.specs().nodes();
+        let mut results = Vec::with_capacity(nodes.len());
+
+        for node in nodes {
+            let mut node_spec = context.registry().specs().guarded_node(node.id()).await?;
+            results.push(delete_unknown_volume_snapshot_reconciler(context, &mut node_spec).await);
+        }
+        Self::squash_results(results)
+    }
+
+    async fn poll_timer(&mut self, _context: &PollContext) -> bool {
+        self.counter.poll()
+    }
+}
+
+/// Identify unknown snapshots and attempt to clean them up.
+#[tracing::instrument(skip(node_spec, context), level = "trace", fields(request.reconcile = true))]
+async fn delete_unknown_volume_snapshot_reconciler(
+    context: &PollContext,
+    node_spec: &mut OperationGuardArc<NodeSpec>,
+) -> PollResult {
+    let node_wrapper = context.registry().node_wrapper(node_spec.uid()).await?;
+
+    if let Ok(snaps) = node_wrapper
+        .list_repl_snapshots(&ListReplicaSnapshots::All)
+        .await
+    {
+        for snap in snaps {
+            if let Some((snap_uuid, txn_id)) = extract_uuid_txn_id(snap.snap_name()) {
+                // Don't cleanup if the snapshot is known, i.e part of present snapshot's
+                // transaction.
+                if let Ok(snap_guard) = context.registry().specs().volume_snapshot(&snap_uuid).await
```

Comment:
There's a bug here, if the guard is taken we ignore and destroy, which is wrong

Reply:
Yeah. Do we need the guard? I think we can use the rsc only.

Reply:
We *do* need the guard, but we can do one of two:
1. delete on not_found error, skip/continue on any other error
2. get rsc first, then try to upgrade to guard and skip/continue on error